src/core/sphinx/peel.rs

```diff
@@ -0,0 +1,190 @@
+		return Err(PeelErr::KxPublic)
+	}
+
+	// (kx_public, mac, encrypted_header, payload) correspond to (alpha, gamma, beta, delta) in the
+	// Sphinx paper
+	let (kx_public, mac, encrypted_header, payload) =
+		array_refs![packet, KX_PUBLIC_SIZE, MAC_SIZE, ENCRYPTED_HEADER_SIZE, PAYLOAD_SIZE];
+
+	let sds = SmallDerivedSecrets::new(kx_shared_secret);
+
+	// Verify the MAC
+	if !mac_ok(mac, encrypted_header, sds.mac_key()) {
+		return Err(PeelErr::Mac)
+	}
+
+	// Decrypt the header and generate padding for length invariance. Try to get the decrypted
+	// header in the right place in the output. The most likely case is that we will be forwarding
+	// the packet to another node, so assume this. We could save some work in the deliver case by
+	// decrypting just the first few bytes to start with to see if we need to decrypt the rest.
+	// This would complicate things and as the forward case is much more common it might ultimately
+	// not make things any faster, so don't bother for now.
+	let decrypted_header = array_mut_ref![
+		out,
+		KX_PUBLIC_SIZE - RAW_ACTION_SIZE,
+		ENCRYPTED_HEADER_SIZE + MAX_HEADER_PAD_SIZE
+	];
+	*array_mut_ref![decrypted_header, 0, ENCRYPTED_HEADER_SIZE] = *encrypted_header;
+	*array_mut_ref![decrypted_header, ENCRYPTED_HEADER_SIZE, MAX_HEADER_PAD_SIZE] =
+		[0; MAX_HEADER_PAD_SIZE]; // Padding is generated by encrypting zeroes
+	apply_header_encryption_keystream(decrypted_header, sds.header_encryption_key());
+
+	let raw_action = RawAction::from_le_bytes(*array_ref![decrypted_header, 0, RAW_ACTION_SIZE]);
+	Ok(match raw_action {
+		RAW_ACTION_DELIVER_REQUEST => {
+			// Peel off the final layer of payload encryption
+			let out = array_mut_ref![out, 0, PAYLOAD_SIZE];
+			*out = *payload;
+			decrypt_payload(out, &derive_payload_encryption_key(kx_shared_secret));
+
+			check_payload_tag(array_ref![out, PAYLOAD_DATA_SIZE, PAYLOAD_TAG_SIZE])?;
+
+			Action::DeliverRequest
+		},
+		RAW_ACTION_DELIVER_REPLY => {
+			// Pull the SURB ID out of the header
+			let surb_id = *array_ref![decrypted_header, RAW_ACTION_SIZE, SURB_ID_SIZE];
+
+			// Copy the payload across but don't do anything with it yet; the caller will need to
+			// fetch the keys corresponding to the SURB ID and then call decrypt_reply_payload()
+			*array_mut_ref![out, 0, PAYLOAD_SIZE] = *payload;
+
+			Action::DeliverReply { surb_id }
+		},
+		RAW_ACTION_DELIVER_COVER => Action::DeliverCover { cover_id: None },
+		RAW_ACTION_DELIVER_COVER_WITH_ID => {
+			// Pull the cover ID out of the header
+			let cover_id = *array_ref![decrypted_header, RAW_ACTION_SIZE, COVER_ID_SIZE];
+
+			Action::DeliverCover { cover_id: Some(cover_id) }
+		},
+		_ => {
+			// Forward. Determine target...
+			let target = if raw_action == RAW_ACTION_FORWARD_TO_PEER_ID {
+				// Copy out peer ID and move rest of header down
+				let peer_id = *array_ref![decrypted_header, RAW_ACTION_SIZE, PEER_ID_SIZE];
+				decrypted_header.copy_within(
```

Comment:
why this copy_within?

Reply:
Maybe it's a bit of a silly micro-optimisation, but the remaining encrypted header ends up in the right place only in the `Target::MixnodeIndex` case; in the `Target::PeerId` case it needs to be shuffled down (as we consume an extra 32 bytes for the peer ID).

Reply:
Ok, I was thinking both packet got same alignment.
Wondering if it can result in some incorrect padding.

Reply:
Ah, the header decryption step also generates the padding, and we conservatively generate the maximum amount of possible padding; some will go unused in the `MixnodeIndex` case.